core/templates/pages/translate-custom-parser.ts

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Custom parser for translations.
+ */
+
+import { TranslateDefaultParser, TranslateParser } from '@ngx-translate/core';
+import constants from 'assets/constants';
+import MessageFormat from 'messageformat';
+import { I18nLanguageCodeService } from 'services/i18n-language-code.service';
+
+export class TranslateCustomParser extends TranslateParser {
+  messageFormat;
+  constructor(
+    private translateDefaultParser: TranslateDefaultParser,
+    private i18nLanguageCodeService: I18nLanguageCodeService
+  ) {
+    super();
+    this.translateDefaultParser.templateMatcher = /<\[\s?([^{}\s]*)\s?\]>/g;
+    this.messageFormat = new MessageFormat();
+  }
+
+  interpolate(
+      expr: string | Function,
+      params?: { [key: string]: number | string | boolean }): string {
+    let interpolate = this.translateDefaultParser.interpolate(expr, params);
+    if (params) {
+      if (params.hasOwnProperty('messageFormat') && interpolate !== undefined) {
+        if (params.messageFormat) {
+          try {
+            interpolate = this.messageFormat.compile(
```

Comment:
```suggestion
            let messageFormat = new MessageFormat();
            interpolate = messageFormat.compile(
```

Can we do this instead of tying messageFormat to this?

Reply:
done

Reply:
Let's not add a property to an object just for testing purposes. Can you find a way to test it without adding messageFormat to parser property?

Reply:
Instantiating MessageFormat inside interpolate function will create a new instance every time a string is interpolated.
If a page has 300 translations then 300 instances of message format will be created.

By instantiating it in the constructor, only 1 instance will be created and used for all translations.

Test Case 1: Inside constructor
![Screenshot from 2021-06-02 13-43-27](https://user-images.githubusercontent.com/24855641/120448510-3f128600-c3a9-11eb-9e75-b992819462d7.png)

Output:

![Screenshot from 2021-06-02 13-43-41](https://user-images.githubusercontent.com/24855641/120448551-4c2f7500-c3a9-11eb-8cdc-13dde9f6b633.png)

Test Case 2: Inside interpolate
![Screenshot from 2021-06-02 13-45-09](https://user-images.githubusercontent.com/24855641/120448586-55b8dd00-c3a9-11eb-9546-a2cc429fbb26.png)

Output:
![Screenshot from 2021-06-02 13-46-00](https://user-images.githubusercontent.com/24855641/120448606-5b162780-c3a9-11eb-8ced-774c4ac1aa40.png)

Reply:
I suggest to keep message format as a parser property.